system/core/database.php

```diff
@@ -659,24 +681,86 @@ function lockAndRead() {
 			Write the given string to the file, replacing any
 			current contents.
 
-			If the file is not opened yet, it is opened and
-			an exclusive lock is taken. After writing, the
-			file is closed and the lock is released.
-
-			If the file was already opened (and thus an
-			exclusive lock is held), the file is not closed
-			after writing, and the lock remains in effect.
-
-			This method can essentially be used just like
-			file_put_contents(), but with locking enabled by
-			default.
+			The file must be already locked before calling
+			this. It is not closed or unlocked after
+			writing.
 		*/
 		function write($content) {
 			if (!$this->fd)
 				throw new LogicException('Cannot write to file ' . $this->filename . ', not locked');
-			ftruncate($this->fd, 0);
-			rewind($this->fd);
-			fwrite($this->fd, $content);
+
+			// Clear the last error, to prevent using an
+			// older error when something goes wrong without
+			// generating a PHP error (e.g. like read/write
+			// before PHP 7.4).
+			error_clear_last();
+
+			// Write data to a tempfile first and then
+			// rename the tempfile over the existing file. This:
+			//  - Ensures that file reading can be done
+			//    without locks, since the actual file will
+			//    never be modified.
+			//  - Prevents corruption on full disk (since
+			//    rename will be atomic, unlike the
+			//    truncate-and-write that would happen when
+			//    writing to the file directly).
+			//
+			// This uses a fixed temp filename instead of
+			// using tempnam, since the latter might default
+			// to a different directory on permission
+			// problems, which would break atomicity if it
+			// is on another filesystem.
+			//
+			$tmpfilename = $this->filename . '.tmp_for_writing';
```

Comment:
is het een idee om hier een `uniqid()` achteraan te plakken?

Reply:
Heb ik over nagedacht, maar in principe zou het lock op de file zelf moeten zorgen dat er nooit twee tegelijk naar dezelfde file schrijven. Mocht er toch ooit een keer een tmpfile blijven rondslingeren heeft *geen* uniqid toevoegen als voordeel dat die tmpfile vanzelf opgeruimd wordt bij een volgende write. Klinkt dat zinnig?

Reply:
Klinkt zinnig :-)